api/envoy/api/v2/cds.proto

```diff
@@ -526,7 +526,7 @@ message Cluster {
   // :ref:`statistics <config_cluster_manager_cluster_stats>` if :ref:`alt_stat_name
   // <envoy_api_field_Cluster.alt_stat_name>` is not provided.
   // Any ``:`` in the cluster name will be converted to ``_`` when emitting statistics.
-  string name = 1 [(validate.rules).string = {min_bytes: 1}];
+  string name = 1 [(validate.rules).string = {min_bytes: 1, pattern: "^[^\0\n\r]+$"}];
```

Comment:
drive by: it doesn't thrill me that we have this non-trivial regex copied all over the place. Is there any way that we could reference the value and then have comments?

Reply:
Ohhh, wow yeah that's pretty awful. I guess there's two options, wrap a string field with validation in a message `ValidHeaderString` to use in places where we need valid header strings, or define some kind of global (does this kind of thing exist?)

Reply:
For the former that would unfortunately be an API breaking change. I guess we could do this for V3 somehow? For the latter I don't know if that is possible. @htuch @lizan?

Reply:
PGV could support something like this if it allowed named patterns, where the {name: value} mapping is supplied at validation time. That's a bigger project. Something spidey sensey suggests CEL could also be useful here potentially.

For v3 could create a base type for strings with this validation applied. It should be possible with some kind of type lifting annotation.

Reply:
Hmmm. Should I hold off on v2 and make this change in to a base string type in v3alpha? 
What do you mean by type lifting annotation? Do you mean that I can leverage an annotation to validate these strings that lifts the message up to (string->validHeaderString) and back to (validHeaderString -> string)? Why not only change over string types here to validHeaderString types?

Reply:
I think you could introduce the type in v2 with alternative fields and deprecate the extant fields immediately, that way it will become the standard way of specifying these fields in v3. This is how we're approaching all API work in v3, there is no magical period at which we define the v3 API, instead we have the v3alpha candidate machine generated from v2, and the work requiring manual restructurings happens in v2 in a non-breaking way.

I'm not sure if it makes sense to hold off, we should try and get useful fuzz stuff landed right away. I'd suggest doing both the new fields with the header string type _and_ adding the boiler plate that is in the PR. The boiler plate fields will be tagged `deprecated`, so I'm comfortable with the tech debt.

The lifting annotation idea was about guiding the v3 proto generation process. I think it's probably overkill TBH, but the idea was to allow a type change in the field as we go from v2 to v3. It's hard to do in general, but we were thinking of doing this for WKTs.

Reply:
In an ideal world we'd use Http::HeaderUtility::headerIsValid for anything which modifies headers.  If we can't do it in header parsing, we could still do it on config ingress, config application, and/or have asserts in the header map functions.

Reply:
Hm. Right now the asserts in header map functions are causing the failure (caught by fuzzers), and it's pretty unclear these are just invalid config values.. 

There's 3 different string values that cause this invalid header issue:

* test.fuzz.Headers - used for populating fuzz test headers, probably should be validated with PGV annotations
* config strings like cluster name or domains later used as header values. Probably should be checked with `headerIsValid` in prod code. 
* config strings that are clearly header keys/values like `internal_only_headers`, `headers_to_add`, `cluster_header` - I personally think these should be a PGV validated `validHeaderKey` or `validHeaderValue` message, but WDYT?

Reply:
One possible argument against doing this in proto validators is that the as the API evolves these validators would need to be set in more places.  It also ties us to protobuf for all APIs, but I guess config has fully moved to protos these days and no direct load from json remains?

An argument for proto validators is that they help with API documentation.  I think it would be an improvement to have some ability to name these patterns things like "HeaderKeyValidator" that verifies that the characters in a config field that contains a header key are valid lowercase token characters.

Reply:
The APIs are canonically protobuf, we are committed to this for sure. My opinion as @envoyproxy/api-shepherds hat wearer is that APIs should ideally be self-documenting and express constraints explicitly rather than implicitly. This argues for PGV annotations here. These annotations can be (in theory at least) consumed by other systems, e.g. other config generation and validation pipelines. This will become increasingly more important in a multi-client world when we have UDPA.

I think we need to do something about reducing boiler plate. I'd argue for better base types (which include these validators in their definition) across the API surface. With the Envoy stable versioning story, we now have a path for forward migration of the API, so I think we should take advantage of it.

Reply:
Hmm. So now at this point, we have `valid_*` fields for many of the (but there must be more missing) messages. I think 
* I may need to tackle these PR in many small PRs per route/cluster/fuzztests etc pieces in order to modify the corresponding code to take `valid_x` instead of `x` if specified.
* if it's a matter of just introducing a named pattern in PGV vs handling these new types, this is starting to look easier
* base types, I agree, are useful, and since we have stable versioning, we should add them.
* I used `Http::HeaderUtility::headerIsValid` where I could
* Yes, these validators do need to be set in more places. If there's base types, people need to remember to use base types. The good thing about base-types is that we can change the validation in one place.